### DISCUSSION TITLE
Access page instance from UserControl

### DISCUSSION THREAD

**Discussion by mnesarco** [flet-dev/flet#845] [Q&A]
Hi Friends,
Is there a way to access the current `page` instance from an `UserControl` without the need to pass it explicitly in constructor? I Want to implement something similar to InheritedWidget/Provider/Hooks but I need access to the root to maintain a pseudo singleton registry.

**Comment by FeodorFitsner** [comment]
Inside `UserControl` implementation it's just `self.page` which is getting set on `self.did_mount()`: https://flet.dev/docs/guides/python/user-controls/

**Comment by mnesarco** [comment]
Inside my UserControl, self.page is always == None

**Reply by mnesarco** [reply]
Do you mean it is self.page is None inside build() but available later after did_mount?

**Comment by mnesarco** [comment]
Ok, I confirmed self.page is None in build, but available in did_mount. It means that I need to build partially in build method and complete in did_mount. It is a bit weird.

**Reply by FeodorFitsner** [reply]
What are you trying to achieve - could you please elaborate?

**Reply by mnesarco** [reply]
Don't worry, this is not an issue but a discussion. I have been working with Flutter and React and may others for years and I find Flutter's and React's hooks quite useful. I have been experimenting with Flet just two days and I love it. Now I am implementing my own framework on top of Flet so I can make data centric web applications quite fast for intranet use. As Flet is still very new and I am just learning it I am trying to understand all the limitations and the architecture.

As the page instance  is the "current session", I want to use it as the application state holder for the current user. The idea is to implement something similar to React Hooks for UserControls.

**Comment by mnesarco** [comment]
Hi @FeodorFitsner ,
I have achieved my goal. Demo:

```python
import fletii as ii
import flet as ft
import random

# Services ------------------------------

# Super complex service interface
class MyService:
    def get_some_data(self, arg):
        pass

# Alternative A
class MyServiceA(MyService):
    def get_some_data(self, arg):
        return f"Hello Mr. {arg}!!!"

# Alternative B
class MyServiceB(MyService):
    def get_some_data(self, arg):
        return f"Bye Mr. {arg}!!!"

# Widgets (Controls) --------------------

# Directly depends on MyService
class MyGreeter(ii.CustomControl):
    def __init__(self, value):
        super().__init__()
        self.value = value

    def build(self, context: ii.BuildContext):
        hw = context.find(MyService) # HERE: Access to scoped service instance
        return ft.Text(hw.get_some_data(self.value))
        
# Transitively depends on MyService
class MyCtl(ii.CustomControl):
    def build(self, context: ii.BuildContext):
        return ft.Column([
            MyGreeter("Paul"),
            MyGreeter("Ringo"),
            MyGreeter("John"),
            MyGreeter("George"),
        ])

# Application ---------------------------

class MyApp(ii.CustomControl):
    def build(self, context: ii.BuildContext):
        # Configure service implementation at runtime...
        context.set_service(MyServiceA() if random.random() > 0.5 else MyServiceB(), key=MyService)
        # Use MyCtl (implementation agnostic)
        return MyCtl()

# Bootstrap -----------------------------

def main(page: ft.Page):
    page.add(MyApp())

ft.app(target=main)
```

- **fletii** is my WIP flet library.
- I can have injectable scoped services, so no need to drill down parameters
- No global singletons

What do you think about this approach? Implementing ii.CustomControl and ii.BuildContext were just 70 lines of code, flet is fantastic. My only concern is how stable are the current apis. I implemented ii.CustomControl based on flet.UserControl.
